api/android/api/jni/nnstreamer-jni.c

```diff
@@ -0,0 +1,321 @@
+
+  (*env)->DeleteLocalRef (env, cls_arraylist);
+  (*env)->DeleteLocalRef (env, obj_arraylist);
+  (*env)->DeleteLocalRef (env, cls_info_list);
+  return TRUE;
+}
+
+/**
+ * @brief Native method for single-shot API.
+ */
+jlong
+Java_com_samsung_android_nnstreamer_Single_nativeOpen (JNIEnv * env, jobject thiz,
+    jstring model, jobject in, jobject out)
+{
+  ml_single_h single;
+  ml_tensors_info_h in_info, out_info;
+  int status;
+  const char *model_info = (*env)->GetStringUTFChars (env, model, NULL);
+
+  single = in_info = out_info = NULL;
+
+  if (in) {
+    ml_tensors_info_create (&in_info);
+    nns_parse_tensors_info (env, in, (ml_tensors_info_s *) in_info);
+  }
+
+  if (out) {
+    ml_tensors_info_create (&out_info);
+    nns_parse_tensors_info (env, out, (ml_tensors_info_s *) out_info);
+  }
+
+  /* supposed tensorflow-lite only for android */
+  status = ml_single_open (&single, model_info, in_info, out_info,
+      ML_NNFW_TYPE_ANY, ML_NNFW_HW_AUTO);
+  if (status != ML_ERROR_NONE) {
+    nns_loge ("Failed to create the pipeline.");
+    single = NULL;
+  }
+
+  (*env)->ReleaseStringUTFChars (env, model, model_info);
+
+  ml_tensors_info_destroy (in_info);
+  ml_tensors_info_destroy (out_info);
+
+  return CAST_TO_LONG (single);
+}
+
+/**
+ * @brief Native method for single-shot API.
+ */
+void
+Java_com_samsung_android_nnstreamer_Single_nativeClose (JNIEnv * env, jobject thiz,
+    jlong handle)
+{
+  ml_single_h single;
+
+  single = CAST_TO_TYPE (handle, ml_single_h);
+
+  ml_single_close (single);
+}
+
+/**
+ * @brief Native method for single-shot API.
+ */
+jobject
+Java_com_samsung_android_nnstreamer_Single_nativeInvoke (JNIEnv * env, jobject thiz,
+    jlong handle, jobject in)
+{
+  ml_single_h single;
+  ml_tensors_data_s *input, *output;
+  int status;
+  jobject result = NULL;
+
+  single = CAST_TO_TYPE (handle, ml_single_h);
+  output = NULL;
+
+  input = g_new0 (ml_tensors_data_s, 1);
+  g_assert (input);
+
+  if (!nns_parse_tensors_data (env, in, input)) {
+    nns_loge ("Failed to parse input data.");
+    goto done;
+  }
+
+  status = ml_single_invoke (single, input, (ml_tensors_data_h *) &output);
+  if (status != ML_ERROR_NONE) {
+    nns_loge ("Failed to get the result from pipeline.");
+    goto done;
+  }
+
+  if (!nns_convert_tensors_data (env, output, &result)) {
+    nns_loge ("Failed to convert the result to data.");
+    result = NULL;
+  }
+
+done:
+  g_free (input);
+  g_free (output);
+  return result;
+}
+
+/**
+ * @brief Native method to initialize NNStreamer.
+ */
+jboolean
+Java_com_samsung_android_nnstreamer_NNStreamer_nativeInitialize (JNIEnv * env, jclass clazz,
+    jobject context)
+{
+  nns_logi ("Called native initialize.");
+
+  if (!gst_is_initialized ()) {
+    nns_loge ("GStreamer is not initialized.");
+    return JNI_FALSE;
+  }
+
+  /* register nnstreamer plugins */
+  GST_PLUGIN_STATIC_REGISTER (nnstreamer);
+
+  /* filter tensorflow-lite sub-plugin */
+  init_filter_tflite ();
```

Comment:
initialize decoder?

Reply:
I added all decoder sub-plugins.

Reply:
Yes. I know. But I don't think it's enough. As far as I know, one of the reason we make decoder sub-plugin is to give free will to the developer so that they can add the post processor what they want. If they have there own decoder, then they have to add init function here and recompile. Am I right?